### PR TITLE
Groups refacto to implément members logic, with members implémentation  

### PR DESCRIPTION
**Pull request by maxime-dodin**
#### Description

Main things in the PR:
Group models and enpoints modification
Member  logic implémentation 

FInally testing for groups and members Request !!! (add firework)

Related Us
#37 , #36, #38, #34, #35, #49 

#### Changelog

- [UPDATE]
 - Group suppression remove all related members

- [ENHANCEMENT]
Members are store in another collection with, 
  -  Member creation function (without internal token logic), we need to discuss about this
  -  Member suppression with two main case ,  if lambda user, or if admin set another random member as Admin
  -  Listing of Groups from an account_id, no need to be same user as the account_id
  -  Listing of members in a group
  - Get Group from groupID

- [TODO]
  - For now i miss update Member information rpc, because exept for the created_at we don't store any information related to the member (RGPD compliance imao)

### CODE REVIEW

File: models/members.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	Group     *string   `json:"group_id" bson:"group_id,omitempty"`
+	Role      string    `json:"role" bson:"role,omitempty"`
+	CreatedAt time.Time `json:"created_at" bson:"created_at,omitempty"`
+}
+
+type MemberPayload struct {
+	Account   *string   `json:"account_id" bson:"account_id,omitempty"`
+	Group     *string   `json:"group_id" bson:"group_id,omitempty"`
+	Role      string    `json:"role" bson:"role,omitempty"`
+	CreatedAt time.Time `json:"created_at" bson:"created_at,omitempty"`
```
Comment (rojas-diego) [inline models/members.go:20]: This should be generated automatically and not a parameter to the `Create` method.

File: models/groups.go [hunk 2]
```diff
@@ -2,30 +2,24 @@ package models
...
+	CreatedAt   time.Time `json:"created_at" bson:"created_at,omitempty"`
 }
 
 type OneGroupFilter struct {
-	ID      string `json:"id" bson:"_id,omitempty"`
-	OwnerID string `json:"owner_id" bson:"owner_id,omitempty"`
+	ID string `json:"id" bson:"_id,omitempty"`
 }
 
 type ManyGroupsFilter struct {
```
Comment (rojas-diego) [inline models/groups.go:25]: Is this intentionally empty for now?

File: models/mongo/members.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+		coll:   db.Collection("members"),
+	}
+
+	_, err := rep.coll.Indexes().CreateOne(
+		context.Background(),
+		mongo.IndexModel{
+			Keys:    bson.D{{Key: "account_id", Value: 1}, {Key: "group_id", Value: 1}},
+			Options: options.Index().SetUnique(true),
+		},
+	)
```
Comment (rojas-diego) [inline models/mongo/members.go:35]: Just curious, is this operation idempotent?

**Review by rojas-diego — COMMENTED** [review]
Neat! I left a few comments. LMK what you think.

File: models/groups.go (hunk 2 above)
Comment (maxime-dodin) [reply models/groups.go:24]: yep, it's unused for now

File: models/mongo/members.go (hunk 3 above)
Comment (maxime-dodin) [reply models/mongo/members.go:35]: i think so, mongo does not allowed dupplicate index, so it's only created once

File: models/groups.go [hunk 4]
```diff
@@ -2,30 +2,24 @@ package models
...
-	Description *string        `json:"description" bson:"description,omitempty"`
-	OwnerID     string         `json:"owner_id" bson:"owner_id,omitempty"`
-	Members     *[]GroupMember `json:"members" bson:"members,omitempty"`
-}
-
-type GroupMember struct {
-	ID string `json:"account_id" bson:"account_id,omitempty"`
+	Name        *string   `json:"name" bson:"name,omitempty"`
+	Description *string   `json:"description" bson:"description,omitempty"`
+	CreatedAt   time.Time `json:"created_at" bson:"created_at,omitempty"`
```
[the remark this answers is not in the document]
Comment (maxime-dodin) [reply models/groups.go]: Fix, it's generate by DB now

File: models/members.go (hunk 1 above)
Comment (maxime-dodin) [reply models/members.go]: Fix, it's generate by DB now

File: models/members.go [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"context"
+	"time"
+)
+
+type Member struct {
+	ID        string    `json:"_id" bson:"_id,omitempty"`
+	Account   *string   `json:"account_id" bson:"account_id,omitempty"`
+	Group     *string   `json:"group_id" bson:"group_id,omitempty"`
```
[the remark this answers is not in the document]
Comment (maxime-dodin) [reply models/members.go]: i have changed the models

File: models/members.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+type MemberPayload struct {
+	Account   *string   `json:"account_id" bson:"account_id,omitempty"`
+	Group     *string   `json:"group_id" bson:"group_id,omitempty"`
+	Role      string    `json:"role" bson:"role,omitempty"`
+	CreatedAt time.Time `json:"created_at" bson:"created_at,omitempty"`
+}
+
+type MemberFilter struct {
+	Account *string `json:"account_id" bson:"account_id,omitempty"`
+	Group   *string `json:"group_id" bson:"group_id,omitempty"`
```
[the remark this answers is not in the document]
Comment (maxime-dodin) [reply models/members.go]: i have changed the models

File: accounts.go [hunk 7]
```diff
@@ -152,13 +152,13 @@ func (srv *accountsAPI) DeleteAccount(ctx context.Context, in *accountsv1.Delete
...
 
-func (srv *accountsAPI) ListAccount(ctx context.Context, in *accountsv1.ListAccountRequest) (*accountsv1.ListAccountResponse, error) {
+func (srv *accountsAPI) ListAccount(ctx context.Context, in *accountsv1.ListAccountsRequest) (*accountsv1.ListAccountsResponse, error) {
 	_, err := srv.authenticate(ctx)
 	if err != nil {
 		return nil, status.Error(codes.Unauthenticated, err.Error())
 	}
 
-	accounts, err := srv.repo.List(ctx, &models.ManyAccountsFilter{}, &models.Pagination{Offset: in.Paginate.Offset, Limit: in.Paginate.Limit})
+	accounts, err := srv.repo.List(ctx, &models.ManyAccountsFilter{}, &models.Pagination{Offset: int64(in.Offset), Limit: int64(in.Limit)})
```
[the remark this answers is not in the document]
Comment (maxime-dodin) [reply accounts.go]: it was missing in the list accounts and getGroup, so I added them

File: groups.go [hunk 8]
```diff
@@ -6,43 +6,53 @@ import (
...
 
 func (srv *groupsAPI) CreateGroup(ctx context.Context, in *accountsv1.CreateGroupRequest) (*accountsv1.CreateGroupResponse, error) {
 	token, err := srv.authenticate(ctx)
 	if err != nil {
-		return nil, err
+		return nil, status.Error(codes.Unauthenticated, err.Error())
 	}
 
-	id := token.UserID.String()
+	account_id := token.UserID.String()
```
[the remark this answers is not in the document]
Comment (maxime-dodin) [reply groups.go]: done,

File: models/members.go [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+	DeleteMany(ctx context.Context, filter *MemberFilter) error
+
+	Get(ctx context.Context, filter *MemberFilter) (*Member, error)
+
+	Update(ctx context.Context, filter *MemberFilter, account *MemberPayload) (*Member, error)
+
+	List(ctx context.Context, filter *MemberFilter) ([]Member, error)
+
+	SetAdmin(ctx context.Context, filter *MemberFilter) error
```
[the remark this answers is not in the document]
Comment (maxime-dodin) [reply models/members.go]: you are right, i updated the methods

**Review by rojas-diego — APPROVED** [review]
Neat! Thank you v much!

**Pull request merged**